app/main.py

```diff
@@ -1,214 +1,61 @@
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
-
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
-
-    # -------------------------------------------------------------------------------
-    # BATTLE:
-
-    # 1 Lancelot vs Mordred:
-    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
-    mordred["hp"] -= lancelot["power"] - mordred["protection"]
-
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
-
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
-
-    # 2 Arthur vs Red Knight:
-    arthur["hp"] -= red_knight["power"] - arthur["protection"]
-    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
-
-    # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
-
-    if red_knight["hp"] <= 0:
-        red_knight["hp"] = 0
-
-    # Return battle results:
-    return {
-        lancelot["name"]: lancelot["hp"],
-        arthur["name"]: arthur["hp"],
-        mordred["name"]: mordred["hp"],
-        red_knight["name"]: red_knight["hp"],
-    }
-
-
-print(battle(KNIGHTS))
+class Knight:
+    def __init__(self,
+                 name: str,
+                 power: int,
+                 hp: int,
+                 protection: int = 0) -> None:
+        self.name = name
+        self.power = power
+        self.hp = hp
+        self.protection = protection
+
+    def add_armour(self, armours: list) -> None:
+        for armour in armours:
+            self.protection += armour["protection"]
+
+    def add_weapon(self, weapon: dict) -> None:
+        self.power += weapon["power"]
+
+    def add_potion(self, potion: dict) -> None:
+        if "power" in potion["effect"]:
+            self.power += potion["effect"]["power"]
+        if "hp" in potion["effect"]:
+            self.hp += potion["effect"]["hp"]
+        if "protection" in potion["effect"]:
+            self.protection += potion["effect"]["protection"]
+
+
+def create_knight(knight: dict) -> Knight:
+    result = Knight(
+        name=knight["name"],
+        power=knight["power"],
+        hp=knight["hp"]
+    )
+    if knight["armour"]:
+        result.add_armour(knight["armour"])
+
+    result.add_weapon(knight["weapon"])
+
+    if knight["potion"]:
+        result.add_potion(knight["potion"])
+    return result
+
+
+def fight(knight: Knight, enemy: Knight) -> None:
+    knight.hp -= enemy.power - knight.protection
+
+    if knight.hp < 0:
+        knight.hp = 0
+
+
+def battle(knights_dict: dict) -> dict:
+    knights = [
+        create_knight(knight=knight[1])
+        for knight in knights_dict.items()
+        if knight[0] is not None
+    ]
+    for index in range(len(knights) - 2):
+        fight(knights[index], knights[index + 2])
+        fight(knights[index + 2], knights[index])
+
+    return {knight.name: knight.hp for knight in knights}
```

Comment:
You should satisfy the following requirements from the task description:

- Split functionality to different modules and packages as you think will be the most meaningful.

Change the project structure to achieve that.